File: backend/app/routers/monitor.py
from __future__ import annotations

from datetime import timedelta

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies.database import get_session
from app.models.models import User, VisitorTrack
from app.routers.admin import get_admin_user
from app.schemas.response import APIResponse

router = APIRouter(prefix="/status", tags=["monitor-status"])

DEFAULT_DAYS = 7
MAX_DAYS = 90


@router.get("/overview")
async def get_overview(
    days: int = Query(DEFAULT_DAYS, ge=1, le=MAX_DAYS, description="统计天数"),
    current_user: User = Depends(get_admin_user),
    session: AsyncSession = Depends(get_session),
):
    """Get visitor overview data (admin only)"""
    from datetime import UTC, datetime

    end_time = datetime.now(UTC)
    start_time = end_time - timedelta(days=days)

    # 1. Total visits
    result = await session.execute(
        select(func.count(VisitorTrack.id)).where(
            VisitorTrack.visit_time >= start_time
        )
    )
    total_visits = result.scalar() or 0

    # 2. Unique visitors (by IP)
    result = await session.execute(
        select(func.count(func.distinct(VisitorTrack.ip_address))).where(
            VisitorTrack.visit_time >= start_time
        )
    )
    unique_visitors = result.scalar() or 0

    # 3. Unique visitors (by visitor_id)
    result = await session.execute(
        select(func.count(func.distinct(VisitorTrack.visitor_id))).where(
            VisitorTrack.visit_time >= start_time
        )
    )
    unique_visitor_ids = result.scalar() or 0

    # 4. Top pages
    result = await session.execute(
        select(
            VisitorTrack.page_path,
            func.count(VisitorTrack.id).label("count"),
        )
        .where(VisitorTrack.visit_time >= start_time)
        .group_by(VisitorTrack.page_path)
        .order_by(func.count(VisitorTrack.id).desc())
        .limit(10)
    )
    top_pages = [
        {"page_path": row[0], "count": row[1]} for row in result.fetchall()
    ]

    # 5. Browser stats
    result = await session.execute(
        select(
            VisitorTrack.browser,
            func.count(VisitorTrack.id).label("count"),
        )
        .where(VisitorTrack.visit_time >= start_time)
        .group_by(VisitorTrack.browser)
        .order_by(func.count(VisitorTrack.id).desc())
    )
    browser_stats = [
        {"browser": row[0], "count": row[1]} for row in result.fetchall()
    ]

    # 6. Daily trend
    result = await session.execute(
        select(
            func.date(VisitorTrack.visit_time).label("date"),
            func.count(VisitorTrack.id).label("count"),
        )
        .where(VisitorTrack.visit_time >= start_time)
        .group_by(func.date(VisitorTrack.visit_time))
        .order_by(func.date(VisitorTrack.visit_time).desc())
    )
    daily_trend = [
        {"date": str(row[0]), "count": row[1]} for row in result.fetchall()
    ]

    return APIResponse.ok(
        data={
            "total_visits": total_visits,
            "unique_visitors": unique_visitors,
            "unique_visitor_ids": unique_visitor_ids,
            "top_pages": top_pages,
            "browser_stats": browser_stats,
            "daily_trend": daily_trend,
            "period_days": days,
        },
        message="Visitor overview data retrieved successfully",
    )


@router.get("/visitors")
async def get_visitors(
    days: int = Query(DEFAULT_DAYS, ge=1, le=MAX_DAYS, description="统计天数"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    current_user: User = Depends(get_admin_user),
    session: AsyncSession = Depends(get_session),
):
    """Get visitor list with pagination (admin only)"""
    from datetime import UTC, datetime

    end_time = datetime.now(UTC)
    start_time = end_time - timedelta(days=days)

    # Get total count
    result = await session.execute(
        select(func.count(VisitorTrack.id)).where(
            VisitorTrack.visit_time >= start_time
        )
    )
    total = result.scalar() or 0

    # Pagination
    offset = (page - 1) * page_size

    # Get list
    result = await session.execute(
        select(VisitorTrack)
        .where(VisitorTrack.visit_time >= start_time)
        .order_by(VisitorTrack.visit_time.desc())
        .offset(offset)
        .limit(page_size)
    )
    visitors = result.scalars().all()

    visitor_list = [
        {
            "id": v.id,
            "visitor_id": v.visitor_id,
            "page_url": v.page_url,
            "page_path": v.page_path,
            "referrer": v.referrer,
            "browser": v.browser,
            "screen_resolution": v.screen_resolution,
            "language": v.language,
            "ip_address": v.ip_address,
            "visit_time": v.visit_time.isoformat() if v.visit_time else None,
        }
        for v in visitors
    ]

    return APIResponse.ok(
        data={
            "list": visitor_list,
            "total": total,
            "page": page,
            "page_size": page_size,
            "total_pages": (total + page_size - 1) // page_size,
        },
        message="Visitor list retrieved successfully",
    )


@router.get("/user-logins")
async def get_user_logins(
    days: int = Query(DEFAULT_DAYS, ge=1, le=MAX_DAYS, description="统计天数"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    current_user: User = Depends(get_admin_user),
    session: AsyncSession = Depends(get_session),
):
    """Get user login logs with pagination (admin only)"""
    from datetime import UTC, datetime

    end_time = datetime.now(UTC)
    start_time = end_time - timedelta(days=days)

    # Get users with login records
    result = await session.execute(
        select(User).where(User.login_count > 0).order_by(User.id.desc())
    )
    users = result.scalars().all()

    # Filter users with recent logins
    login_logs = []
    for user in users:
        has_recent_login = False
        if user.current_login_at and user.current_login_at >= start_time:
            has_recent_login = True
        if user.last_login_at and user.last_login_at >= start_time:
            has_recent_login = True

        if has_recent_login:
            login_logs.append(
                {
                    "user_id": user.id,
                    "username": user.username,
                    "name": user.name,
                    "login_count": user.login_count,
                    "last_login_at": (
                        user.last_login_at.isoformat()
                        if user.last_login_at
                        else None
                    ),
                    "current_login_at": (
                        user.current_login_at.isoformat()
                        if user.current_login_at
                        else None
                    ),
                    "last_login_ip": user.last_login_ip,
                    "current_login_ip": user.current_login_ip,
                }
            )

    # Pagination
    total = len(login_logs)
    offset = (page - 1) * page_size
    paginated_logs = login_logs[offset : offset + page_size]

    return APIResponse.ok(
        data={
            "list": paginated_logs,
            "total": total,
            "page": page,
            "page_size": page_size,
            "total_pages": (total + page_size - 1) // page_size,
        },
        message="User login logs retrieved successfully",
    )

File: go-backend/internal/dto/monitor.go
package dto

import "time"

// Overview 对齐 Python monitor_service.get_overview 的返回结构。
type Overview struct {
	TotalVisits      int                      `json:"total_visits"`
	UniqueVisitors   int                      `json:"unique_visitors"`
	UniqueVisitorIDs int                      `json:"unique_visitor_ids"`
	TopPages         []map[string]any         `json:"top_pages"`
	BrowserStats     []map[string]any         `json:"browser_stats"`
	OSSStats         []map[string]any         `json:"os_stats"`
	DailyTrend       []map[string]any         `json:"daily_trend"`
	PeriodDays       int                      `json:"period_days"`
}

// VisitorItem 单条访客记录，字段名与 Python visit 字典对齐。
type VisitorItem struct {
	ID               uint       `json:"id"`
	VisitorID        string     `json:"visitor_id"`
	PageURL          string     `json:"page_url"`
	PagePath         string     `json:"page_path"`
	Referrer         *string    `json:"referrer"`
	Browser          *string    `json:"browser"`
	ScreenResolution *string    `json:"screen_resolution"`
	Language         *string    `json:"language"`
	IPAddress        string     `json:"ip_address"`
	VisitTime        *time.Time `json:"visit_time"`
}

// Visitors 分页访客列表。
type Visitors struct {
	List       []VisitorItem `json:"list"`
	Total      int           `json:"total"`
	Page       int           `json:"page"`
	PageSize   int           `json:"page_size"`
	TotalPages int           `json:"total_pages"`
}

// UserLoginItem 单条用户登录记录，字段名与 Python login_logs 对齐。
type UserLoginItem struct {
	UserID         int     `json:"user_id"`
	Username       string  `json:"username"`
	Name           string  `json:"name"`
	LoginCount     int     `json:"login_count"`
	LastLoginAt    *string `json:"last_login_at"`
	CurrentLoginAt *string `json:"current_login_at"`
	LastLoginIP    *string `json:"last_login_ip"`
	CurrentLoginIP *string `json:"current_login_ip"`
	Active         bool    `json:"active"`
}

// UserLogins 分页用户登录记录。
type UserLogins struct {
	List       []UserLoginItem `json:"list"`
	Total      int             `json:"total"`
	Page       int             `json:"page"`
	PageSize   int             `json:"page_size"`
	TotalPages int             `json:"total_pages"`
}

// ServerStatus 对齐 Python _get_server_status_payload 字段名。
type ServerStatus struct {
	CPUPercent float64 `json:"cpu_percent"`
	CPUCores  int     `json:"cpu_cores"`
	MemTotal  int     `json:"mem_total"`
	MemUsed   int     `json:"mem_used"`
	MemUsage  float64 `json:"mem_usage"`
	DiskTotal float64 `json:"disk_total"`
	DiskUsed  float64 `json:"disk_used"`
	DiskUsage float64 `json:"disk_usage"`
}
